### PR TITLE
ENV (os-release) & CSV file parsing (CFE-1881 ENT-3043 ENT-3046)

### PR DESCRIPTION
**Pull request by olehermanse**
Thanks to @tzz for #2624.

### CODE REVIEW

**Comment by olehermanse** [comment]
Not working yet. Need to write unit tests and debug some more.

**Comment by tzz** [comment]
Nice cleanups @olehermanse

**Comment by olehermanse** [comment]
@jimis Do yo agree with the files and dependencies as they are now?

File: libutils/Makefile.am [hunk 1]
```diff
@@ -47,6 +47,8 @@ libutils_la_SOURCES = \
 	proc_keyvalue.c proc_keyvalue.h \
 	bool.h \
 	json.c json.h json-priv.h \
+	csv-json.c csv-json.h \
+	env-json.c env-json.h \
```
Comment (jimis) [inline libutils/Makefile.am]: Creating a separate function for each file is a bit excessive. Either add them to `json.[ch]` or create something like `json_utils.[ch]` for all JSON utility functions.

**Comment by jimis** [comment]
The removal of `EvalContext` from libenv is admittedly too ambitious so forget that, but the cleaning up of evalfunction.c from the various parsing functionality was much needed. I like what I see, so keep it up! Of course +1 for the unit tests as always!

**Comment by olehermanse** [comment]
Started in jenkins: https://ci.cfengine.com/job/bootstrap-enterprise-pr/176/

File: libutils/Makefile.am (hunk 1 above)
Comment (olehermanse) [inline libutils/Makefile.am]: Fixed.

**Comment by olehermanse** [comment]
@jimis Review this at your leisure. I'm thinking this is such a big PR now that it makes sense to do the CoreOS class in a follow up PR.

**Comment by olehermanse** [comment]
@nick.anderson can you take a look at the acceptance test and see if you agree with the input env vs expected json output?

**Comment by nickanderson** [comment]
@olehermanse Nothing jumps out at me as being incorrect.

I do not see any single quote or unfinished quotes in the example data though:

KEY='value'
KEY="#oops forgot to close it

**Comment by olehermanse** [comment]
@nickanderson Are single quotes used anywhere?

**Comment by olehermanse** [comment]
@nickanderson Single quotes and unclosed quotes are now supported, and added to both unit and acceptance tests.

**Comment by olehermanse** [comment]
Started in jenkins.

@jimis @kacf Please review.

**Comment by olehermanse** [comment]
Started in jenkins: https://ci.cfengine.com/job/bootstrap-enterprise-pr/185/

File: libutils/json-utils.c [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    *value_out = NULL;
+
+    char *line = TrimWhitespace(raw_line);
+    if (NULL_OR_EMPTY(line))
+    {
+        return;
+    }
+
+    const char *myname = "ParseEnvLine";
+    int line_length = strlen(line);
```
Comment (jimis) [inline libutils/json-utils.c]: `size_t`

File: libutils/json-utils.c [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+    if (line[0] == '#' || line_length == 0)
+    {
+        return;
+    }
+    char *next_equal = strchr(line, '=');
+    if (next_equal == NULL)
+    {
+        Log(LOG_LEVEL_DEBUG, "%s: The line '%s' in ENV file '%s' isn't empty, but was skipped because it's missing an equal sign",
```
Comment (jimis) [inline libutils/json-utils.c]: Don't log the whole line. It could be a bogus binary file. :-) Try something like "While parsing ENV file %s, skipping line %d because ...".

File: libutils/json-utils.c [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            myname, line, filename_for_log);
+        return;
+    }
+    int equal_index = (int)(next_equal - line);
```
Comment (jimis) [inline libutils/json-utils.c]: There is a reason you got a warning, and it's bad that your cast silenced it. :-)

Proper type is `size_t` or `long` or even `intptr_t`.

File: libutils/json-utils.c [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    if (equal_index == 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "%s: The line '%s' in ENV file '%s' was skipped because it's missing a key",
+            myname, line, filename_for_log);
+        return;
+    }
+    *next_equal = '\0';
+    key = TrimWhitespace(line);
+    value = TrimWhitespace(next_equal + 1);
+    value = filtered_copy(value, value);
```
Comment (jimis) [inline libutils/json-utils.c]: :-o

File: libutils/json-utils.c [hunk 6]
```diff
@@ -0,0 +1,263 @@
+#include <json-utils.h>
+#include <logging.h>    // Log()
+#include <file_lib.h>   // safe_fopen()
+#include <string_lib.h> // TrimWhitespace()
+#include <csv_parser.h>
+#include <alloc.h>
+#define CF_BUFSIZE 4096
```
Comment (jimis) [inline libutils/json-utils.c]: don't redefine an existing define. Maybe create your own define, or just use the number you want if you use it once.

File: libutils/json-utils.c [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+/**
```
Comment (jimis) [inline libutils/json-utils.c]: Nice helper functions in this file, which could (should?) be unit-checked very nicely. But how,  if they are internal?

FYI a common method for unit testing `static` functions is including the `.c` file on top of the unit test file. :-) For reasons, this is even better many times than exporting the function in a header file and linking to the whole library.

File: libutils/file_lib.h [hunk 8]
```diff
@@ -125,5 +125,10 @@ bool FileSparseClose(int fd, const char *filename,
                      size_t total_bytes_written,
                      bool last_write_was_hole);
 
+/**
+ * @brief Works exactly like posix 'getline', EXCEPT it does not include carriage return at the end.
+ * @return -1 on error OR EOF, so check. Or bytes in buff without null terminator.
```
Comment (olehermanse) [inline libutils/file_lib.h]: Context @jimis, context! :) 
> return -1 on error OR EOF, so check. Or bytes in buff without null terminator.
the return value is number of bytes in buffer, EXCLUDING null terminator.

File: libutils/json-utils.c [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                ++src;
+                if (*src == opening_quote)
+                {
+                    break;
+                }
+            }
+        }
+        if (*src == '\\')
+        {
+            ++src;
```
Comment (olehermanse) [inline libutils/json-utils.c]: See tests for examples :) `\n` is not supported. `\"` and `\'` will add those characters to buffer instead of interpreting them as quoted strings. `\\` will of course add a backslash. In fact, for simplicity backslash-anything will just add that character to the value buffer.

Comment (olehermanse) [inline libutils/json-utils.c]: The reason I added them was because @tzz had some in his tests, and I figured they might be useful if people want to use this env file format for simple key value store.

File: libutils/json-utils.c [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    while (CfReadLine(&raw_line, &line_size, fin) != -1)
+    {
+        ++linenumber;
+        int line_length = strlen(raw_line);
+        byte_count += line_length;
+        if (byte_count > size_max)
+        {
+            ByteLimitLog(myname, linenumber, input_path, size_max);
+            break;
```
Comment (olehermanse) [inline libutils/json-utils.c]: I believe this was there in @tzz original PR. The logic is that the byte count limit is for the whole file, not just one line. Although I don't see a reason why it has to be like this.

File: libutils/json-utils.c (hunk 9 above)
Comment (jimis) [inline libutils/json-utils.c]: No problem, just add some description and TODO comments in the function docstring so that we know what's missing. Deferring to the tests is not really handy.

File: libutils/file_lib.h (hunk 8 above)
Comment (jimis) [inline libutils/file_lib.h]: then rephrase as "number of bytes in buff excluding ..." :-)

File: libutils/json-utils.c (hunk 10 above)
Comment (tzz) [inline libutils/json-utils.c]: I'm OK with either strategy (line limit or file limit). Typically at the user level (policy) we let them control the file limit but default to `inf` (e.g. see `readjson()`). But the module protocol, for instance, does line limits.

File: libutils/json-utils.c (hunk 5 above)
Comment (olehermanse) [inline libutils/json-utils.c]: What? This?
```
value = filtered_copy(value, value);
```
Copy from value to value, returning the new starting pos in value(dest) :)

File: libutils/json-utils.c [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    *value_out = value;
+}
+
+/**
+ * @brief Opens an env file, prints an error message on failure
+ *
+ * @param input_path Path to env file
+ * @param myname Descriptive name of caller for logging purposes
+ */
+static FILE *ParseEnvOpenFile(const char *input_path, const char *myname)
```
Comment (olehermanse) [inline libutils/json-utils.c]: I started making the os-release class function, and that function will use these, which is why I separated them out.

File: libutils/json-utils.c [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+    if (fin == NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s cannot open the ENV file '%s' (fopen: %s)",
+            myname, input_path, GetErrorStr());
+
+    }
+    return fin;
+}
+
+static void ByteLimitLog(const char *myname, int linenumber, const char *input_path, size_t size_max)
```
Comment (olehermanse) [inline libutils/json-utils.c]: Same as last: I started making the os-release class function, and that function will use these, which is why I separated them out.

File: libutils/json-utils.c [hunk 13]
```diff
@@ -0,0 +1,263 @@
+#include <json-utils.h>
```
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c (hunk 10 above)
Comment (olehermanse) [inline libutils/json-utils.c]: @jimis Any preference?

File: libutils/json-utils.c (hunk 6 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Changed to `ENV_BYTE_LIMIT`

File: libutils/json-utils.c (hunk 10 above)
Comment (jimis) [inline libutils/json-utils.c]: I misunderstood the code then. I'm OK with both, so if it makes your life easier keep the total byte count limit. And the log message being hidden in a function didn't help when reading that code.

File: libutils/json-utils.c (hunk 7 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Added unit tests for `filtered_copy` !

File: libutils/json-utils.c (hunk 12 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+    assert(dst);
+    char *beginning = dst;
+    char opening_quote = '\0';
+    while (*src != '\0')
+    {
+        if (*src == '\"' || *src == '\'')
+        {
+            if (opening_quote == *src)
+            {
+                break;
```
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            }
+            else if (opening_quote == '\0')
+            {
+                opening_quote = *src;
+                ++src;
+                if (*src == opening_quote)
```
Comment (olehermanse) [inline libutils/json-utils.c]: Rewrote to detect some invalid input strings like `AB"CD`

File: libutils/json-utils.c (hunk 9 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Added some notes on backslashes and what's supported

File: libutils/json-utils.c (hunk 2 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c (hunk 3 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c (hunk 4 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c (hunk 5 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Added a descriptive comment.

File: libutils/json-utils.c (hunk 11 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Removed separate function for now.

File: libutils/json-utils.c [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+
+/**
+ * @brief Parses an env file and creates a key-value pair json element
+ *
+ * Creates JSON element where all keys and values are strings
+ *
+ * @param input_path file to read from ex: "/etc/os-release"
+ * @param size_max   Maximum size of env file (in bytes)
+ * @param json_out   Where to save pointer to new JsonElement, must destroy
+ * @return 0 = success
```
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+ */
+int JsonParseEnvFile(const char *input_path, size_t size_max, JsonElement **json_out)
+{
+    assert(json_out);
+    *json_out = JsonObjectCreate(10);
+    const char *myname = "JsonParseEnvFile";
+    size_t line_size = CF_BUFSIZE;
+    char *raw_line = xmalloc(CF_BUFSIZE);
```
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+    FILE *fin = ParseEnvOpenFile(input_path, myname);
+    if (fin == NULL)
+    {
+        return -1;
+    }
+
+    while (CfReadLine(&raw_line, &line_size, fin) != -1)
+    {
+        ++linenumber;
+        int line_length = strlen(raw_line);
```
Comment (olehermanse) [inline libutils/json-utils.c]: Fixed

File: libutils/json-utils.c (hunk 10 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Moved log message and keeping the file size limit

File: libutils/json-utils.c [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+        byte_count += line_length;
+        if (byte_count > size_max)
+        {
+            ByteLimitLog(myname, linenumber, input_path, size_max);
+            break;
+        }
+
+        ParseEnvLine(raw_line, &key, &value, input_path);
+        if (key != NULL && value != NULL)
```
Comment (olehermanse) [inline libutils/json-utils.c]: Added some log messages to ParseEnvLine, for invalid syntax, missing key, etc. Will not log on all cases of NULL - I don't want to fill the log with debug messages about empty lines or commented out lines.

File: libutils/file_lib.h (hunk 8 above)
Comment (olehermanse) [inline libutils/file_lib.h]: Fixed

File: libpromises/evalfunction.c [hunk 20]
```diff
@@ -6390,110 +6461,23 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
...
-            Log(LOG_LEVEL_VERBOSE, "%s cannot open the CSV file '%s' (fopen: %s)",
-                fp->name, input_path, GetErrorStr());
-            return FnFailure();
-        }
-
-        JsonElement *json = JsonArrayCreate(50);
-        int linenumber = 0;
-        char *line;
-        while ((line = GetCsvLineNext(fin)) != NULL)
+        else if (0 == strcmp(fp->name, "readenvfile"))
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: Fixed

File: libutils/json-utils.c [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+            myname, input_path, GetErrorStr());
+        JsonDestroy(*json_out);
+        return -1;
+    }
+
+    free(raw_line);
+    return 0;
+}
+
+int JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **json_out)
```
Comment (olehermanse) [inline libutils/json-utils.c]: This is code from Ted's pull request

File: libpromises/evalfunction.c [hunk 22]
```diff
@@ -6340,6 +6341,72 @@ static FnCallResult FnCallReadRealList(EvalContext *ctx, ARG_UNUSED const Policy
     return ReadList(ctx, fp, args, CF_DATA_TYPE_REAL);
 }
 
+JsonElement *ReadDataFile(const char *log_identifier, const char* input_path,
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: Renamed and moved to `json-utils`

File: libpromises/evalfunction.c [hunk 23]
```diff
@@ -6390,110 +6461,23 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
...
-                JsonArrayAppendArray(json, line_arr);
-            }
+            requested_mode = "ENV";
         }
-
-        bool atend = feof(fin);
-        fclose(fin);
-
-        if (!atend)
+        else
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: This is not code I've written. I think it's clear as is, but I can try to add some description.

Comment (olehermanse) [inline libpromises/evalfunction.c]: Added this comment further up in the function
```
    // JsonReadDataFile reads key-value pairs from a json, yaml, csv or
    // env file and puts them in a JSON element
    // requested_mode is used to specify the input file type
    // requested_mode is set based on file extension or function name
    // example: readyaml -> "YAML" , ".csv" -> "CSV" etc.
    // defaults to "JSON"
```
I cannot see how I would comment on the specific else. It is just one line, defaulting to JSON if all other comparisons fail.

**Comment by olehermanse** [comment]
Created a separate branch and PR for the memory leaks I'm fixing: #2800

**Comment by olehermanse** [comment]
Started in jenkins: https://ci.cfengine.com/job/bootstrap-enterprise-pr/195/

@jimis Fixed all the requests. Added coreos class and flavor. Ready to merge AFAIK, waiting for jenkins and travis.

**Comment by olehermanse** [comment]
I should test this on a coreos machine on thursday.

File: libenv/sysinfo.c [hunk 24]
```diff
@@ -1085,14 +1085,24 @@ static void OSClasses(EvalContext *ctx)
 
     if (os_release_json != NULL)
     {
-        EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "os_release", os_release_json, CF_DATA_TYPE_CONTAINER, "source=agent,os,release,derived_from=/etc/os-release");
+        EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS, "os_release",
+                                      os_release_json, CF_DATA_TYPE_CONTAINER,
+                                      "source=agent,os,release,derived_from=/etc/os-release");
```
Comment (jimis) [inline libenv/sysinfo.c]: I believe it's `derived-from` with dash, not underscore. Also you'd better use `derived-from-file`. Pinging @tzz to verify, he is the tag master. ;-)

File: libenv/sysinfo.c [hunk 25]
```diff
@@ -1085,14 +1085,24 @@ static void OSClasses(EvalContext *ctx)
...
         const char *os_release_id = JsonObjectGetAsString(os_release_json, "ID");
-        if(os_release_id != NULL)
+        const char *os_release_version = JsonObjectGetAsString(os_release_json, "VERSION");
+        char os_release_flavor[256];
+        snprintf(os_release_flavor, 256, "%s_%s", os_release_id, os_release_version);
```
Comment (jimis) [inline libenv/sysinfo.c]: What if they are NULL?

File: libenv/sysinfo.c [hunk 26]
```diff
@@ -1085,14 +1085,24 @@ static void OSClasses(EvalContext *ctx)
...
+        char *dot_pointer = os_release_flavor;
+        while( (dot_pointer = strchr(dot_pointer, '.')) != NULL )
         {
-            if(strcmp(os_release_id, "coreos") == 0)
-            {
-                EvalContextClassPutHard(ctx, "coreos", "inventory,attribute_name=none,source=agent");
-            }
+            *dot_pointer = '_';
```
Comment (jimis) [inline libenv/sysinfo.c]: You probably want `CanonifyNameInPlace()` to account for other characters invalid for classes.

File: libenv/sysinfo.c (hunk 25 above)
Comment (jimis) [inline libenv/sysinfo.c]: Also see `xasprintf()` for an allocating printf.

**Comment by jimis** [comment]
No worries, but it will take more time now to review the whole code again. Please append small commits from now on.
> Sorry! Didn't see this:
>> Please don't rebase this pull request, just add more commits as you fix things so that we can keep track. In the end we can squash and rebase everything.

**Comment by olehermanse** [comment]
Fixed error in enterprise and restarted in jenkins:
Started in jenkins: https://ci.cfengine.com/job/bootstrap-enterprise-pr/202/

**Comment by jimis** [comment]
I hate github sometimes; after the rebase it's very hard to follow up on your fixes.

/me thinks we should use other tools for code review, there are a bunch of them.

File: libenv/sysinfo.c (hunk 24 above)
Comment (olehermanse) [inline libenv/sysinfo.c]: Changed to `derived-from-file`

File: libenv/sysinfo.c (hunk 25 above)
Comment (olehermanse) [inline libenv/sysinfo.c]: If they are NULL, we cannot access the os-release file. We still do all the other detection code, but won't be able to determine and set coreos class.

File: libenv/sysinfo.c (hunk 24 above)
Comment (tzz) [inline libenv/sysinfo.c]: YES

File: libenv/sysinfo.c (hunk 25 above)
Comment (olehermanse) [inline libenv/sysinfo.c]: Changed to xasprintf()

File: libenv/sysinfo.c (hunk 26 above)
Comment (olehermanse) [inline libenv/sysinfo.c]: @jimis This is not a class, but a variable. Does this still apply?

File: libpromises/evalfunction.c (hunk 23 above)
Comment (jimis) [inline libpromises/evalfunction.c]: Thanks, it's now more clear; the complexity isn't trivial since `FnCallReadData()` is called from policy functions `readdata()`, `readcsv()`, `readjson()`, `readyaml()`, `readenvfile()`. Mode of operation is modified according to function name and `requested_mode` parameter.

File: libenv/sysinfo.c (hunk 26 above)
Comment (jimis) [inline libenv/sysinfo.c]: I thought you were setting classes. I believe it applies, since these variable values are canonified class-like. Don't ask me why.

Plus you'll probably have to set classes as well for full CoreOS support. Here are the classes on Centos 7:
```
centos                                                       inventory,attribute_name=none,source=agent,hardclass
centos_7                                                     inventory,attribute_name=none,source=agent,hardclass
centos_7_3                                                   inventory,attribute_name=none,source=agent,hardclass
```

**Comment by olehermanse** [comment]
@jimis This is ready AFAIK. It's pending in jenkins and running in travis atm. Tell me when you want me to squash the commits.

**Comment by olehermanse** [comment]
In jenkins: https://ci.cfengine.com/job/bootstrap-enterprise-pr/204/

**Comment by olehermanse** [comment]
@jimis Squashed and waiting for jenkins.

**Comment by jimis** [comment]
jenkins is green

**Comment by olehermanse** [comment]
> jenkins is green
@jimis For this PR? Nice. So this is ready to merge AFAIK (?). One more review of whole PR (?).

**Comment by jimis** [comment]
yes

File: libenv/sysinfo.c [hunk 27]
```diff
@@ -982,6 +984,55 @@ static void SetFlavour(EvalContext *ctx, const char *flavour)
...
+                }
+                free(os_release_version);
+                free(os_release_flavor);
+            }
+            free(os_release_id);
+        }
+        free(tags);
+        JsonDestroy(os_release_json);
+    }
+}
```
Comment (jimis) [inline libenv/sysinfo.c]: +1 nicely written! Resource allocation and destruction is well structured.

File: libutils/json-utils.c (hunk 13 above)
Comment (jimis) [inline libutils/json-utils.c]: Just for the sake of completion, in libpromises you can also include `cf3.defs.h` on top, this way you get more standard defines and it auto-includes `platform.h`.

File: tests/unit/env_file_test.c [hunk 28]
```diff
@@ -0,0 +1,186 @@
+#include <test.h>
+#include <json-utils.c>
```
Comment (jimis) [inline tests/unit/env_file_test.c]: This is a good way to unit-test, but please don't do it so subtly, in-between normal `.h` includes. Somebody who reads the code must notice that you are including a `.c` file here, so make it stand out a bit.

File: libutils/json-utils.c (hunk 13 above)
Comment (olehermanse) [inline libutils/json-utils.c]: Okay, noted

**Comment by olehermanse** [comment]
@jimis Merge once travis passes

**Comment by olehermanse** [comment]
@jimis Please merge :)

**Comment by nickanderson** [comment]
Yay

**Comment by jimis** [comment]
Awesome work @olehermanse and @tzz ! HOORAY!

**Comment by tzz** [comment]
I can't wait to use this. Thanks!
